staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -397,10 +398,15 @@ func NewCacherFromConfig(config Config) (*Cacher, error) {
 		// so that future reuse does not get a spurious timeout.
 		<-cacher.timer.C
 	}
-	progressRequester := newConditionalProgressRequester(config.Storage.RequestWatchProgress, config.Clock)
+	// Add grpc context metadata to watch and progress notify requests done by cacher to:
+	// * Prevent starvation of watch opened by cacher, by moving it to separate Watch RPC than watch request that bypass cacher.
+	// * Ensure that progress notification requests are executed on the same Watch RPC as their watch, which is required for it to work.
+	contextMetadata := metadata.New(map[string]string{"source": "watch-cache"})
```

Comment:
do normal non-watch requests (like gets/writes/deletes) mix in the same RPC with misbehaving client watches that can cause problems, or only other watch requests?

Reply:
Watch is assigned to [RCP by a key](https://github.com/kubernetes/kubernetes/blob/d2b4928669c633cffb0e4aa6317d0e016ee37de6/vendor/go.etcd.io/etcd/client/v3/watch.go#L322) defined as [grpc metadata from context](https://github.com/kubernetes/kubernetes/blob/d2b4928669c633cffb0e4aa6317d0e016ee37de6/vendor/go.etcd.io/etcd/client/v3/watch.go#L1037-L1042). 

Assuming that watch cache uses a unique grpc metadata it will create a dedicated RPC stream for it.